src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5154,3 +5162,75 @@ void MacroAssembler::verify_cross_modify_fence_not_required() {
   }
 }
 #endif
+
+
+// Stack frame creation/removal
+
+void MacroAssembler::enter()
+{
+  protect_return_address();
+  stp(rfp, lr, Address(pre(sp, -2 * wordSize)));
+  mov(rfp, sp);
+}
+
+void MacroAssembler::leave()
+{
+  mov(sp, rfp);
+  ldp(rfp, lr, Address(post(sp, 2 * wordSize)));
+  authenticate_return_address();
+}
+
+// ROP Protection
+
+void MacroAssembler::protect_return_address() {
```

Comment:
We need proper, full, detailed comments about what these functions do, with reference to primary AArch64 documentation.

Reply:
As far as the AArch64 docs are concerned the relevant details are provided in ARM-ARM D

- The PAC functionality is described in ARM-ARM Section D5.1.5
- Overview of the PAC instructions is provided in section C3.1.9
- Detailed PAC instruction descriptions are provided in C6.2.195 - C6.2.199

n.b. I am specifically referring to my (possibly out of date) copy ARM-DDI 0487D.a (ID103018) which is the Initial v8.4 EAC release from 2018.

That said, I agree that a description of how these functions use the underlying PAC support and what, effectively, they achieve via that usage would be necessary. A reference to the relevant sections of the ARM doc in the code would be helpful.

Reply:
Correction:
Using the most up to date ARM ARM G  [ARM DDI 0487G.a (ID011921)]

- The PAC functionality is described in ARM-ARM Section D5.1.5
- Overview of the PAC instructions is provided in section C3.1.10
- Detailed PAC instruction descriptions are provided in C6.2.208 - C6.2.212

Reply:
I'm thinking for references to the Arm Arm to use header titles instead of section numbers, as the titles should be more stable.

Also probably need some description around the code in the pauth_aarch64.hpp too. But I want to make sure I'm not duplicating comments - maybe the macroassembler comments should point to the pauth_aarch64 comments.

It didn't seen common in the code to describe instruction functionality, which is why I didn't add any. Agreed it needs something added though.

Reply:
Yeah. At the definitions of `authenticate_return_address()` et al you can say what you expect in the normal case and what you expect when you've been hacked, along with an overview. I realize that it was a bit tricky to make this work with HotSpot because we're synthesizing return addresses just like hackers do, so a comment where we're patching return addresses would be nice.

As long as the instructions are easily findable in the docs that's good.

Reply:
Just to be clear: no, don't describe instructions. describe what the macros do, and when to use them. Imagine that you, the reader can't see the contents of the macro at all, just the name and the comments.